developer_guides/firmware/component-tutorial/tut-iii-runtime-params.rst

```diff
@@ -0,0 +1,212 @@
+                   return -EINVAL;
+           }
+
+           if (cdata->data->size != sizeof(cd->channel_volume)) {
+                   trace_amp_error("amp_cmd_set_data() error: "
+                                   "invalid data size %d",
+                                   cdata->data->size);
+                   return -EINVAL;
+           }
+
+           memcpy_s(cd->channel_volume, sizeof(cd->channel_volume),
+                    cdata->data->data, cdata->data->size);
+           trace_amp("Amplifier new settings vol[0] %d vol[1] %d",
+                     cd->channel_volume[0], cd->channel_volume[1]);
+           return 0;
+   }
+
+Add another one to report parameters back to the host. Note how the
+``cdata->data`` (``struct sof_abi_hdr``) is updated.
+
+.. code-block:: c
+
+   static int amp_cmd_get_data(struct comp_dev *dev,
+                               struct sof_ipc_ctrl_data *cdata, int max_size)
+   {
+           struct amp_comp_data *cd = comp_get_drvdata(dev);
+
+           if (cdata->cmd != SOF_CTRL_CMD_BINARY) {
+                   trace_amp_error("amp_cmd_get_data() error: invalid cmd %d",
+                                   cdata->cmd);
+                   return -EINVAL;
+           }
+
+           if (sizeof(cd->channel_volume) > max_size)
+                   return -EINVAL;
+
+           memcpy_s(cdata->data->data,
+                    ((struct sof_abi_hdr *)(cdata->data))->size,
+                    cd->channel_volume,
+                    sizeof(cd->channel_volume));
+           cdata->data->abi = SOF_ABI_VERSION;
+           cdata->data->size = sizeof(cd->channel_volume);
+
+           return 0;
+   }
+
+Put everything together as a command handler.
+
+.. code-block:: c
+
+   static int amp_cmd(struct comp_dev *dev, int cmd, void *data, int max_data_size)
+   {
+           struct sof_ipc_ctrl_data *cdata = data;
+           int ret = 0;
+
+           switch (cmd) {
+           case COMP_CMD_SET_DATA:
+                   ret = amp_cmd_set_data(dev, cdata);
+                   break;
+           case COMP_CMD_GET_DATA:
+                   ret = amp_cmd_get_data(dev, cdata, max_data_size);
+                   break;
+           default:
+                   trace_amp_error("amp_cmd() error: unhandled command %d", cmd);
+                   ret = -EINVAL;
+                   break;
+           }
+           return ret;
+   }
+
+And attach the handler to your component driver API.
+
+.. code-block:: c
+   :emphasize-lines: 7
+
+   struct comp_driver comp_amp = {
+           .type = SOF_COMP_AMP,
+           .ops = {
+                   .new = amp_new,
+                   .free = amp_free,
+                   .params = NULL,
+                   .cmd = amp_cmd,
+                   .trigger = amp_trigger,
+                   .prepare = amp_prepare,
+                   .reset = amp_reset,
+                   .copy = amp_copy,
+                   .cache = NULL
+           },
+   };
+
+
+Binary Bytes KControl in Topology
```

Comment:
IMO KControl  may be moved to "topology & SW" section

Reply:
Yes, this content will be "translated" into .m4 modification guide. I saved the content of private bytes here since they will be included in one of the .m4 files.